man/skel.8
.TH SKEL 8 "Sat Dec 23 2006" "" "Network UPS Tools (NUT)"
.\" This line is a comment.
.\"
.\" When copying this man page for your new driver, be sure to replace all
.\" occurrances of "skel" and "SKEL" by your actual driver name.
.\"
.\" Another option is to embed the man page information into your driver source
.\" code.  See the tripplite_usb.c file for one way of doing this, using Perl's
.\" pod2man translator. The advantage of this is that you can keep the
.\" documentation near the source code, which may help remind you to update the
.\" documentation after changing the driver.
.\"
.\" On most systems, you can test the formatting of your man page by running:
.\"   man ./skel.8
.\" (the dot-slash tells man not to search its usual path.)
.\"
.\" After writing a man page, be sure to add it to the appropriate variable in
.\" Makefile.am in this directory.
.\"
.\" On many systems, you can find a reference to the man macros by typing
.\" 'man 7 man'.
.\"
.\" In the "NAME" section, you must follow the format below, including
.\" separating the driver name from its description by "\-". This is to ensure
.\" that the apropos(8) database is properly rebuilt.
.SH NAME
skel \- skeleton driver man page

.SH NOTE
This man page only documents the hardware\(hyspecific features of the
\fBskel\fR driver.  For information about the core driver, see
\fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
.B skel
supports ...

.\" If the driver only works with certain cables, this is a good place to
.\" mention it:
.\" .SH CABLING

.SH EXTRA ARGUMENTS
This driver also supports the following optional settings:

.IP "option1=\fInum\fR"
Set the value of ... to \fInum\fR. Contrast with \fBoption2\fR.
 
.IP "option2=\fIstring\fR"
Some other option.

.\" Optional: list supported instant commands here:
.\" .SH INSTANT COMMANDS

.\" Optional: use INSTALLATION if you need special parameters to the configure
.\" script.
.\" .SH INSTALLATION

.\" You may leave this as "none known at this time", or describe any trouble
.\" encountered when implementing the protocol for your UPS.
.SH KNOWN ISSUES AND BUGS
.SS "Got EPERM: Operation not permitted upon driver startup"

You have forgotten to install the hotplug files, as explained
in the INSTALLATION section above. Don't forget to restart
hotplug so that it applies these changes.

.\" An email address is not strictly necessary, but you may wish to provide
.\" some form of contact information so that users can report bugs.
.SH AUTHORS
J. Random User <user@example.org>

.\" If this driver is ever made obsolete by another driver, mention the
.\" replacement driver in the "SEE ALSO" section. You may also wish to point
.\" the user to other drivers which may better support their hardware, if there
.\" is ambiguity based on the driver name.
.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.5 - what's new since 2.0.4:

 This release is a backport of the development version.  Many changes
 have already been backported previously.  Thus it is more a
 synchronisation release, though it includes many bugfixes and support
 for new models.

 - support for new devices: APC Smart-UPS with 6TI firmware; Belkin
   Small Enterprise F6C1500-TW-RK; Compaq R3000 XR, R5500 XR; Cyber
   Power 550SL, 725SL, 685AVR, 800AVR, 1200AVR, AE550, BC1200 and
   many similar models; Eltek; Inform GUARD; Microsol Rhino; Opti-UPS
   PowerES 420E; PowerMan RealSmart, BackPro; Powerware PW9315 3-phase;
   SOLA 305; Tripp-Lite SMART550USB, SMART2200RMXL2U, OMNI1000LCD,
   OMNI900LCD, OMNI650LCD, 1500 LCD, AVR550U; Viewsonic PowerES 420E.

 - bcmxcp: added 3-phase support

 - megatec: better hardware support, more instant commands

 - mge-hid: support more instant commands

 - newhidups: fixed APC and Tripp Lite bugs, various memory bugs,
   improved report buffering, improved Solaris support, added
   '-x explore' option for easy diagnosis of new devices

 - solis: shutdown programming, support new cables, Solaris support

 - tripplite_usb: updated SMARTPRO support, fixed OL/OB reporting, 
   better error handling, some memory bugs

 - new dummy-ups driver simulator

 - added HTML interface for access to CGI scripts

 - powerpanel: new driver for CyberPower BC1200 and similar models
 
 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.4 - what's new since 2.0.3:

 - The newhidups critical bug (segmentation fault) has been fixed. It has
   also received some more care, like buxfixes and new models support and
   enhancement for Solaris.
   [Peter Selinger and Arnaud Quette]

 - A bug has been fixed in NUT core to support resuming from suspend-to-disk.
   This should also fix other similar issues, like time synchronisation
   through the NTP - Network Time Protocol.
   [Arjen de Korte]

 - The mge-shut driver now better detects the Low Battery status, support
   new models and fixes some wrong status and data. It also fixes some
   issue where the UPS wasn't restarting (refer to mge-shut manpage).
   [Arnaud Quette]

 - The genericups custom configuration through ups.conf is working again
   [Arjen de Korte]

 - The genericups driver type 22 also support CyberPower 725SL
   (and maybe others SL models)
   [David Kaufman]

 - The new megatec driver, which will replace a bunch of drivers by nut 2.2
   (refer to docs/megatec.txt and UPGRADING) has been backported from the
   trunk (Development tree). The powermust driver has also received some
   attention.
   [Carlos Rodrigues]

 - The new rhino driver was added to support Microsol Rhino UPS hardware
   The solis has also been improved for solaris compatibility, and
   internal / external shutdown programming. solis can now save external
   shutdown programming to ups, and support new cables for solis 3
   [Silvino B. MagalhÃ£es]

 - Several fixes and improvements have been made to upsrw, upsset,
   cpsups, tripplite_usb and the FAQ.
   [Arjen de Korte and Charles Lepple]

---------------------------------------------------------------------------
Release notes for NUT 2.0.3 - what's new since 2.0.2:

 - The recent and major newhidups changes have been backported from the
   Development tree. It now:
   - supports models from MGE UPS SYSTEMS, APC and Belkin. Mustek and Unitek
     units are also recognized for development purpose,
   - handles better device reopening, after a disconnection,
   - handles multiple devices, with several parameters to find the right UPS.
   [Peter Selinger, Charles Lepple and Arnaud Quette]

 - The bcmxcp_usb driver has been added to support Powerware USB units.
   [Wolfgang Ocker and Kjell Claesson]

 - The tripplite_usb driver has been added to support Tripp Lite USB units.
   [Charles Lepple]

 - The sec driver is back as gamatronic
   [Gamatronic, Nadav Moskovitch]

 - The genericups driver has received official care from Gamatronic
   to add support for the Gamatronic UPS with alarm interface.
   [Gamatronic, Nadav Moskovitch]

 - The powermust driver now supports Soyntec Sekury C 500 and C 800 units.
   [Hanno Borns]

 - The mge-shut driver has received a bit of attention too, and enhance
   ups.model retrieval for some specific case (release 0.65)

 - The drivers don't change to the "statepath" directory anymore at
   initialisation time if called using -k. This avoid unneeded
   failure to poweroff the UPS if /var is already unmounted.
   [Gaspar Bakos]

 - The belkinunv driver now supports Belkin F6C1100-UNV
   [Dave Breiland]

 - The isbmex driver has been upgraded to version 0.05, which fixes
   various errors in formulas, add shutdown capability and revert
   back baudrate to B9600 (instead of B2400), as it broke the
   communication
   [Ricardo Martinezgarza]

 - The support of Sysgration UPGUARDS Pro650 in fentonups has
   been fixed
   [Simon J. Rowe]

 - The packaging files for Red Hat have received various fixes
   [Thomas Jarosch]

 - The solis driver has been fixed to avoid a naming colision and
   compile on Solaris
   [Paweł Kierdelewicz]

 - The snmp-ups driver has corrected the problem when exposing
   certain time data.

---------------------------------------------------------------------------
Release notes for NUT 2.0.2 - what's new since 2.0.1:

 - the newhidups USB driver has been improved a lot and is no more
   experimental. It also now has a basic APC support, which will
   soon replace the legacy hidups driver.

 - The mge-utalk driver has improved its support for old units.

 - The mge-shut driver has been improved for restart/shutdown
   sequences which was previously blocking the serial port.


drivers/skel.c
/*
	anything commented is optional
	anything else is mandatory
*/

#include "main.h"

/* #include "serial.h" */

/* #define ENDCHAR	'\r'	*/
/* #define IGNCHARS	""	*/

#define DRV_VERSION	"0.01"

void upsdrv_initinfo(void)
{
	/* try to detect the UPS here - call fatal_with_errno() if it fails */

	dstate_setinfo("driver.version.internal", "%s", DRV_VERSION);
	/* dstate_setinfo("ups.mfr", "skel driver"); */
	/* dstate_setinfo("ups.model", "longrun 15000"); */

	/* upsh.instcmd = instcmd; */
}

void upsdrv_updateinfo(void)
{
	/* int flags; */
	/* char temp[256]; */

	/* ser_sendchar(upsfd, 'A'); */
	/* ser_send(upsfd, "foo%d", 1234); */
	/* ser_send_buf(upsfd, bincmd, 12); */

	/* 
	 * ret = ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, IGNCHARS);
	 *
	 * if (ret < STATUS_LEN) {
	 * 	upslogx(LOG_ERR, "Short read from UPS");
	 *	dstate_datastale();
	 *	return;
	 * }
	 */

	/* dstate_setinfo("var.name", ""); */

	/* if (ioctl(upsfd, TIOCMGET, &flags)) {
	 *	upslog_with_errno(LOG_ERR, "TIOCMGET");
	 *	dstate_datastale();
	 *	return;
	 * }
	 */

	/* status_init();
	 *
	 * if (ol)
	 * 	status_set("OL");
	 * else
	 * 	status_set("OB");
	 * ...
	 *
	 * status_commit();
	 *
	 * dstate_dataok();
	 */

	/*
	 * poll_interval = 2;
	 */
}

void upsdrv_shutdown(void)
{
	/* tell the UPS to shut down, then return - DO NOT SLEEP HERE */

	/* maybe try to detect the UPS here, but try a shutdown even if
	   it doesn't respond at first if possible */

	/* replace with a proper shutdown function */
	fatalx("shutdown not supported");

	/* you may have to check the line status since the commands
	   for toggling power are frequently different for OL vs. OB */

	/* OL: this must power cycle the load if possible */

	/* OB: the load must remain off until the power returns */
}

/*
static int instcmd(const char *cmdname, const char *extra)
{
	if (!strcasecmp(cmdname, "test.battery.stop")) {
		ser_send_buf(upsfd, ...);
		return STAT_INSTCMD_HANDLED;
	}

	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
	return STAT_INSTCMD_UNKNOWN;
}
*/

void upsdrv_help(void)
{
}

/* list flags and values that you want to receive via -x */
void upsdrv_makevartable(void)
{
	/* allow '-x xyzzy' */
	/* addvar(VAR_FLAG, "xyzzy", "Enable xyzzy mode"); */

	/* allow '-x foo=<some value>' */
	/* addvar(VAR_VALUE, "foo", "Override foo setting"); */
}

void upsdrv_banner(void)
{
	printf("Network UPS Tools - Skeleton UPS driver %s (%s)\n\n", 
		DRV_VERSION, UPS_VERSION);
}

void upsdrv_initups(void)
{
	/* upsfd = ser_open(device_path); */
	/* ser_set_speed(upsfd, device_path, B1200); */

	/* probe ups type */

	/* to get variables and flags from the command line, use this:
	 *
	 * first populate with upsdrv_buildvartable above, then...
	 *
	 *                   set flag foo : /bin/driver -x foo
	 * set variable 'cable' to '1234' : /bin/driver -x cable=1234
	 *
	 * to test flag foo in your code:
	 *
	 * 	if (testvar("foo"))
	 * 		do_something();
	 *
	 * to show the value of cable:
	 *
	 *      if ((cable = getval("cable")))
	 *		printf("cable is set to %s\n", cable);
	 *	else
	 *		printf("cable is not set!\n");
	 *
	 * don't use NULL pointers - test the return result first!
	 */

	/* the upsh handlers can't be done here, as they get initialized
	 * shortly after upsdrv_initups returns to main.
	 */

	/* don't try to detect the UPS here */
}

void upsdrv_cleanup(void)
{
	/* free(dynamic_mem); */
	/* ser_close(upsfd, device_path); */
}
